File: src/chi_testbench_gem5_v2/sequences/idle.cc
/*
 * Copyright (c) 2026 ripopov
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include "chi_testbench_gem5_v2/driver.hh"
#include "chi_testbench_gem5_v2/sequence_context.hh"
#include "chi_testbench_gem5_v2/sequences/registry.hh"

namespace gem5
{
namespace chi_gem5tb_v2
{
namespace
{

// Idle sequence: no CHI traffic. The fiber returns immediately; the
// driver signals the finish barrier (if present) and stays event-less
// until sim exit.
void
idle_seq(SequenceContext & /*ctx*/)
{}

[[maybe_unused]] Registrar _r("idle", &idle_seq);

} // namespace
} // namespace chi_gem5tb_v2
} // namespace gem5

File: ruby-book/final/chi_testbench_gem5_v2/scenarios/idle.py
# Copyright (c) 2026 ripopov
# SPDX-License-Identifier: BSD-3-Clause
"""
idle — every tile's driver runs the `idle` sequence (no CHI traffic).

Used as the skeleton smoke-test: verifies the v2 build links, SimObjects
instantiate, the mesh + HN-Fs + drivers wire up, and every driver's
fiber enters + returns + signals the barrier cleanly.
"""

from m5.objects import (
    ChiDriverNode,
    ChiGem5V2Barrier,
)


def build(args, planner):
    barrier = ChiGem5V2Barrier(expected=args.num_cpus)

    drivers = []
    for tile in range(args.num_cpus):
        drivers.append(
            ChiDriverNode(
                sequence="idle",
                tile_id=tile,
                finish_barrier=barrier,
            )
        )
    return drivers
